bin/verdicts.py

```diff
@@ -0,0 +1,452 @@
+    with __setitem__, which infers verdicts upwards in the tree as they become
+    available (and retuns the topmost inferred testgroup).
+    Verdicts (registered and inferred) are accessed with __getitem__
+
+    >>> V = Verdicts(["a/b/1", "a/b/2", "a/c/1", "a/d/1", "b/3"])
+    >>> V['a/b/1'] = 'ACCEPTED'
+    >>> V['a/b/2'] = 'AC' # returns 'a/b' because that verdict will be set as well
+    >>> print(V['a/b'], V['.'])
+    ACCEPTED None
+
+    Attributes:
+    - duration[testcase]: the duration of the testcase
+    - children[testgroup]: the set of children of the given testnode
+    - first_error[testgroup]: first child with non-ACCEPTED verdict; None if none exists
+    - first_unknown[testgroup]: first child whose verdict is not (yet) known; None if none exists
+
+    - _verdict[testnode]: the verdict at the given testnode, or None. In particular,
+        verdict['.'] is the root verdict, sometimes called final verdict or submission verdict.
+        Should not be directly set; use __setitem__ on the Verdict object instead.
+    - _unknowns[testgroup]: iterator over the children that do not (yet) have a verdict,
+        in lexicographic order
+    """
+
+    def unknowns_iterator(self, node):
+        """Yield the node's (yet) unknown children in lexicographic order."""
+        for child in sorted(self.children[node]):
+            if self._verdict[child] is not None:
+                continue
+            yield child
+
+    def __init__(self, testcase_list: list[str]):
+        testcases = set(testcase_list)
+        testgroups: set[str] = set(str(path) for tc in testcases for path in Path(tc).parents)
+        self._verdict: dict[str, Verdict | None] = {g: None for g in testcases | testgroups}
+        self.duration: dict[str, int | None] = {g: None for g in testcases}
+
+        self.children: dict[str, set[str]] = {node: set() for node in testgroups}
+        for node in testcases | testgroups:
+            if node != '.':
+                parent = str(Path(node).parent)
+                self.children[parent].add(node)
+        self.first_error: dict[str, str | None] = {node: None for node in testgroups}
+        self.num_unknowns: dict[str, int] = {node: len(self.children[node]) for node in testgroups}
+        self._unknowns = {node: self.unknowns_iterator(node) for node in testgroups}
+        self.first_unknown: dict[str, str | None] = {
+            node: next(self._unknowns[node]) for node in testgroups
+        }
+
+    def is_testgroup(self, node) -> bool:
+        """Is the given testnode name a testgroup (rather than a testcase)?
+        This assumes nonempty testgroups.
+        """
+        return node in self.children
+
+    def is_testcase(self, node) -> bool:
+        """Is the given testnode name a testcase (rather than a testgroup)?
+        This assumes nonempty testgroups.
+        """
+        return node not in self.children
+
+    def __setitem__(self, testcase, verdict: str | Verdict, duration=None):
```

Comment:
Can `__setitem__` even have a third argument? https://github.com/RagnarGrootKoerkamp/BAPCtools/pull/358/files#diff-dcd3fdd39ac1d0938a32c9c9c9ef70fc8b00f9d336d0676e30f159ee0e8d2314R276 doesn't use it either :stuck_out_tongue: